Many multi-value returns

fitzgen opened the pull request:
~~**Do not merge! This is a work-in-progress!**~~ EDIT: ready for review now :)

I'm working on supporting arbitrarily many return values (although we will need to enforce some reasonable upper bound) in cranelift functions, and therefore in Wasm as well.

The approach I've been trying is to:

* Make use of the `sret` special argument for passing in a return pointer, pointing to space allocated in the caller's stack frame for the return values.

* Introduce new `ArgumentLocation` and `ValueLocation` variants for essentially "offset `x` from this function's `sret` parameter": `*(sret + x)`.

* During legalization, we rewrite signatures, calls, and returns to introduce usage of the `sret` return pointer parameter. Specifically,

  * `legalize_signature` will add the `sret` parameter and assign offsets for each return value (currently implemented),

  * `handle_call_abi` will rewrite multi-value calls to allocate a stack slot for the return values, pass a pointer to that space as the `sret` argument, and then load the return values out of the stack slot (not yet implemented),

  * and `handle_return_abi` will rewrite multi-value returns to store the values into the `sret` return pointer at the proper offsets, and return the `sret` return pointer in a register.

This almost works! However, I'm running into a little trouble during register allocation.

Consider this function that returns four `i64`s:

```
function %return_4_i64s() -> i64, i64, i64, i64 fast {
ebb0:
    v0 = iconst.i64 0
    v1 = iconst.i64 1
    v2 = iconst.i64 2
    v3 = iconst.i64 3
    return v0, v1, v2, v3
}
```

After legalization, we see that an `sret` parameter has been added, that all the returns are assigned to offsets from the `sret`, and that we are storing the return values at those offsets just before returning. Great!

```
function %return_4_i64s(i64 sret [%rdi]) -> i64 [sret(0x0)], i64 [sret(0x8)], i64 [sret(0x10)], i64 [sret(0x18)], i64 sret [%rax] fast {
                                ebb0(v4: i64):
[RexOp1pu_id#b8,sret(0)]            v0 = iconst.i64 0
[RexOp1pu_id#b8,sret(8)]            v1 = iconst.i64 1
[RexOp1pu_id#b8,sret(16)]           v2 = iconst.i64 2
[RexOp1pu_id#b8,sret(24)]           v3 = iconst.i64 3
[RexOp1st#8089]                     store notrap aligned v0, v4
[RexOp1stDisp8#8089]                store notrap aligned v1, v4+8
[RexOp1stDisp8#8089]                store notrap aligned v2, v4+16
[RexOp1stDisp8#8089]                store notrap aligned v3, v4+24
[Op1ret#c3]                         return v0, v1, v2, v3, v4
}
```

However, when register allocation comes along, it unnecessarily inserts an instruction to move `v0` from `%rax` to `%rsi`. I'm a bit confused why this is happening, since I update the `ValueLocation` to an offset from the `sret`, and it shouldn't even think that `v0` is live in `%rax`! Anyways, even if this unnecessary instruction were included, things would otherwise work, except that the register allocator also recorded the new location of `v0` as `%rsi`, which means that now the verifier complains that teh arguments are in the wrong places!

```
function %return_4_i64s(i64 sret [%rdi]) -> i64 [sret(0x0)], i64 [sret(0x8)], i64 [sret(0x10)], i64 [sret(0x18)], i64 sret [%rax] fast {
                                ebb0(v4: i64 [%rdi]):
[RexOp1pu_id#b8,%rax]               v0 = iconst.i64 0
[RexOp1pu_id#b8,%rcx]               v1 = iconst.i64 1
[RexOp1pu_id#b8,%rdx]               v2 = iconst.i64 2
[RexOp1pu_id#b8,%rbx]               v3 = iconst.i64 3
[RexOp1st#8089]                     store notrap aligned v0, v4
[RexOp1stDisp8#8089]                store notrap aligned v1, v4+8
[RexOp1stDisp8#8089]                store notrap aligned v2, v4+16
[RexOp1stDisp8#8089]                store notrap aligned v3, v4+24
[RexOp1rmov#8089]                   regmove v0, %rax -> %rsi
[RexOp1rmov#8089]                   regmove v4, %rdi -> %rax
[Op1ret#c3]                         return v0, v1, v2, v3, v4
;~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
; error: inst4: ABI expects v0 at return pointer offset 0, got %rsi

}
```

Any idea what might be going on and what I should do from here?

<details>

<summary>Full debug logs</summary>

```
     Running `target/debug/clif-util compile -dtp /home/fitzgen/cranelift/filetests/wasm/multi-val-return-i64.clif --target x86_64`
 DEBUG cranelift_codegen::timing::details > timing: Starting Parsing textual Cranelift IR, (during <no pass>)
 DEBUG cranelift_codegen::timing::details > timing: Ending Parsing textual Cranelift IR
 DEBUG cranelift_codegen::timing::details > timing: Starting Compilation passes, (during <no pass>)
 DEBUG cranelift_codegen::timing::details > timing: Starting Verify Cranelift IR, (during Compilation passes)
 DEBUG cranelift_codegen::timing::details > timing: Starting Control flow graph, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details > timing: Ending Control flow graph
 DEBUG cranelift_codegen::timing::details > timing: Starting Dominator tree, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details > timing: Ending Dominator tree
 DEBUG cranelift_codegen::timing::details > timing: Starting Verify CPU flags, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details > timing: Ending Verify CPU flags
 DEBUG cranelift_codegen::timing::details > timing: Ending Verify Cranelift IR
 DEBUG cranelift_codegen::context         > Compiling:
function %return_4_i64s() -> i64, i64, i64, i64 fast {
ebb0:
    v0 = iconst.i64 0
    v1 = iconst.i64 1
    v2 = iconst.i64 2
    v3 = iconst.i64 3
    return v0, v1, v2, v3
}

 DEBUG cranelift_codegen::timing::details > timing: Starting Control flow graph, (during Compilation passes)
 DEBUG cranelift_codegen::timing::details > timing: Ending Control flow graph
 DEBUG cranelift_codegen::timing::details > timing: Starting Legalization, (during Compilation passes)
 DEBUG cranelift_codegen::legalizer::boundary > Adding 1 special-purpose arguments to return v0, v1, v2, v3
 DEBUG cranelift_codegen::timing::details     > timing: Ending Legalization
 DEBUG cranelift_codegen::context             > Legalized:
function %return_4_i64s(i64 sret [%rdi]) -> i64 [sret(0x0)], i64 [sret(0x8)], i64 [sret(0x10)], i64 [sret(0x18)], i64 sret [%rax] fast {
                                ebb0(v4: i64):
[RexOp1pu_id#b8,sret(0)]            v0 = iconst.i64 0
[RexOp1pu_id#b8,sret(8)]            v1 = iconst.i64 1
[RexOp1pu_id#b8,sret(16)]           v2 = iconst.i64 2
[RexOp1pu_id#b8,sret(24)]           v3 = iconst.i64 3
[RexOp1st#8089]                     store notrap aligned v0, v4
[RexOp1stDisp8#8089]                store notrap aligned v1, v4+8
[RexOp1stDisp8#8089]                store notrap aligned v2, v4+16
[RexOp1stDisp8#8089]                store notrap aligned v3, v4+24
[Op1ret#c3]                         return v0, v1, v2, v3, v4
}

 DEBUG cranelift_codegen::timing::details     > timing: Starting Verify Cranelift IR, (during Compilation passes)
 DEBUG cranelift_codegen::timing::details     > timing: Starting Control flow graph, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details     > timing: Ending Control flow graph
 DEBUG cranelift_codegen::timing::details     > timing: Starting Dominator tree, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details     > timing: Ending Dominator tree
 DEBUG cranelift_codegen::timing::details     > timing: Starting Verify CPU flags, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details     > timing: Ending Verify CPU flags
 DEBUG cranelift_codegen::timing::details     > timing: Ending Verify Cranelift IR
 DEBUG cranelift_codegen::timing::details     > timing: Starting Dominator tree, (during Compilation passes)
 DEBUG cranelift_codegen::timing::details     > timing: Ending Dominator tree
 DEBUG cranelift_codegen::timing::details     > timing: Starting Remove unreachable blocks, (during Compilation passes)
 DEBUG cranelift_codegen::timing::details     > timing: Ending Remove unreachable blocks
 DEBUG cranelift_codegen::timing::details     > timing: Starting Verify Cranelift IR, (during Compilation passes)
 DEBUG cranelift_codegen::timing::details     > timing: Starting Control flow graph, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details     > timing: Ending Control flow graph
 DEBUG cranelift_codegen::timing::details     > timing: Starting Dominator tree, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details     > timing: Ending Dominator tree
 DEBUG cranelift_codegen::timing::details     > timing: Starting Verify CPU flags, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details     > timing: Ending Verify CPU flags
 DEBUG cranelift_codegen::timing::details     > timing: Ending Verify Cranelift IR
 DEBUG cranelift_codegen::timing::details     > timing: Starting Register allocation, (during Compilation passes)
 DEBUG cranelift_codegen::timing::details     > timing: Starting RA liveness analysis, (during Register allocation)
 DEBUG cranelift_codegen::timing::details     > timing: Ending RA liveness analysis
 DEBUG cranelift_codegen::timing::details     > timing: Starting Verify live ranges, (during Register allocation)
 DEBUG cranelift_codegen::timing::details     > timing: Ending Verify live ranges
 DEBUG cranelift_codegen::timing::details     > timing: Starting RA coalescing CSSA, (during Register allocation)
 DEBUG cranelift_codegen::regalloc::coalescing > Coalescing for:
function %return_4_i64s(i64 sret [%rdi]) -> i64 [sret(0x0)], i64 [sret(0x8)], i64 [sret(0x10)], i64 [sret(0x18)], i64 sret [%rax] fast {
                                ebb0(v4: i64):
[RexOp1pu_id#b8,sret(0)]            v0 = iconst.i64 0
[RexOp1pu_id#b8,sret(8)]            v1 = iconst.i64 1
[RexOp1pu_id#b8,sret(16)]           v2 = iconst.i64 2
[RexOp1pu_id#b8,sret(24)]           v3 = iconst.i64 3
[RexOp1st#8089]                     store notrap aligned v0, v4
[RexOp1stDisp8#8089]                store notrap aligned v1, v4+8
[RexOp1stDisp8#8089]                store notrap aligned v2, v4+16
[RexOp1stDisp8#8089]                store notrap aligned v3, v4+24
[Op1ret#c3]                         return v0, v1, v2, v3, v4
}

 DEBUG cranelift_codegen::regalloc::coalescing > After union-find phase:
 DEBUG cranelift_codegen::timing::details      > timing: Ending RA coalescing CSSA
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify Cranelift IR, (during Register allocation)
 DEBUG cranelift_codegen::timing::details      > timing: Starting Control flow graph, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Control flow graph
 DEBUG cranelift_codegen::timing::details      > timing: Starting Dominator tree, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Dominator tree
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify CPU flags, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify CPU flags
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify Cranelift IR
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify live ranges, (during Register allocation)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify live ranges
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify CSSA, (during Register allocation)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify CSSA
 DEBUG cranelift_codegen::timing::details      > timing: Starting RA spilling, (during Register allocation)
 DEBUG cranelift_codegen::regalloc::spilling   > Spilling for:
function %return_4_i64s(i64 sret [%rdi]) -> i64 [sret(0x0)], i64 [sret(0x8)], i64 [sret(0x10)], i64 [sret(0x18)], i64 sret [%rax] fast {
                                ebb0(v4: i64):
[RexOp1pu_id#b8,sret(0)]            v0 = iconst.i64 0
[RexOp1pu_id#b8,sret(8)]            v1 = iconst.i64 1
[RexOp1pu_id#b8,sret(16)]           v2 = iconst.i64 2
[RexOp1pu_id#b8,sret(24)]           v3 = iconst.i64 3
[RexOp1st#8089]                     store notrap aligned v0, v4
[RexOp1stDisp8#8089]                store notrap aligned v1, v4+8
[RexOp1stDisp8#8089]                store notrap aligned v2, v4+16
[RexOp1stDisp8#8089]                store notrap aligned v3, v4+24
[Op1ret#c3]                         return v0, v1, v2, v3, v4
}

 DEBUG cranelift_codegen::regalloc::spilling   > Spilling ebb0:
 DEBUG cranelift_codegen::regalloc::spilling   > Inst v0 = iconst.i64 0, Pressure[ 1+0/14 0+0/16 ]
 DEBUG cranelift_codegen::regalloc::spilling   > Inst v1 = iconst.i64 1, Pressure[ 2+0/14 0+0/16 ]
 DEBUG cranelift_codegen::regalloc::spilling   > Inst v2 = iconst.i64 2, Pressure[ 3+0/14 0+0/16 ]
 DEBUG cranelift_codegen::regalloc::spilling   > Inst v3 = iconst.i64 3, Pressure[ 4+0/14 0+0/16 ]
 DEBUG cranelift_codegen::regalloc::spilling   > Inst store.i64 notrap aligned v0, v4, Pressure[ 5+0/14 0+0/16 ]
 DEBUG cranelift_codegen::regalloc::spilling   > Inst store.i64 notrap aligned v1, v4+8, Pressure[ 5+0/14 0+0/16 ]
 DEBUG cranelift_codegen::regalloc::spilling   > Inst store.i64 notrap aligned v2, v4+16, Pressure[ 5+0/14 0+0/16 ]
 DEBUG cranelift_codegen::regalloc::spilling   > Inst store.i64 notrap aligned v3, v4+24, Pressure[ 5+0/14 0+0/16 ]
 DEBUG cranelift_codegen::regalloc::spilling   > Inst return v0, v1, v2, v3, v4, Pressure[ 5+0/14 0+0/16 ]
 DEBUG cranelift_codegen::regalloc::spilling   >   reguse: v4@op4/fixed
 DEBUG cranelift_codegen::timing::details      > timing: Ending RA spilling
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify Cranelift IR, (during Register allocation)
 DEBUG cranelift_codegen::timing::details      > timing: Starting Control flow graph, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Control flow graph
 DEBUG cranelift_codegen::timing::details      > timing: Starting Dominator tree, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Dominator tree
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify CPU flags, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify CPU flags
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify Cranelift IR
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify live ranges, (during Register allocation)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify live ranges
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify CSSA, (during Register allocation)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify CSSA
 DEBUG cranelift_codegen::timing::details      > timing: Starting RA reloading, (during Register allocation)
 DEBUG cranelift_codegen::regalloc::reload     > Reload for:
function %return_4_i64s(i64 sret [%rdi]) -> i64 [sret(0x0)], i64 [sret(0x8)], i64 [sret(0x10)], i64 [sret(0x18)], i64 sret [%rax] fast {
                                ebb0(v4: i64):
[RexOp1pu_id#b8,sret(0)]            v0 = iconst.i64 0
[RexOp1pu_id#b8,sret(8)]            v1 = iconst.i64 1
[RexOp1pu_id#b8,sret(16)]           v2 = iconst.i64 2
[RexOp1pu_id#b8,sret(24)]           v3 = iconst.i64 3
[RexOp1st#8089]                     store notrap aligned v0, v4
[RexOp1stDisp8#8089]                store notrap aligned v1, v4+8
[RexOp1stDisp8#8089]                store notrap aligned v2, v4+16
[RexOp1stDisp8#8089]                store notrap aligned v3, v4+24
[Op1ret#c3]                         return v0, v1, v2, v3, v4
}

 DEBUG cranelift_codegen::regalloc::reload     > Reloading ebb0:
 DEBUG cranelift_codegen::timing::details      > timing: Ending RA reloading
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify Cranelift IR, (during Register allocation)
 DEBUG cranelift_codegen::timing::details      > timing: Starting Control flow graph, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Control flow graph
 DEBUG cranelift_codegen::timing::details      > timing: Starting Dominator tree, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Dominator tree
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify CPU flags, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify CPU flags
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify Cranelift IR
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify live ranges, (during Register allocation)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify live ranges
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify CSSA, (during Register allocation)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify CSSA
 DEBUG cranelift_codegen::timing::details      > timing: Starting RA coloring, (during Register allocation)
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring for:
function %return_4_i64s(i64 sret [%rdi]) -> i64 [sret(0x0)], i64 [sret(0x8)], i64 [sret(0x10)], i64 [sret(0x18)], i64 sret [%rax] fast {
                                ebb0(v4: i64):
[RexOp1pu_id#b8,sret(0)]            v0 = iconst.i64 0
[RexOp1pu_id#b8,sret(8)]            v1 = iconst.i64 1
[RexOp1pu_id#b8,sret(16)]           v2 = iconst.i64 2
[RexOp1pu_id#b8,sret(24)]           v3 = iconst.i64 3
[RexOp1st#8089]                     store notrap aligned v0, v4
[RexOp1stDisp8#8089]                store notrap aligned v1, v4+8
[RexOp1stDisp8#8089]                store notrap aligned v2, v4+16
[RexOp1stDisp8#8089]                store notrap aligned v3, v4+24
[Op1ret#c3]                         return v0, v1, v2, v3, v4
}

 DEBUG cranelift_codegen::regalloc::coloring   > Coloring ebb0:
 DEBUG cranelift_codegen::regalloc::coloring   > Start ebb0 with entry-diversion set to
      { }
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring v0 = iconst.i64 0
    from [ GPR: acdb--s-89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     glob [ GPR: acdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     color v0 -> %rax
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring v1 = iconst.i64 1
    from [ GPR: -cdb--s-89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     glob [ GPR: acdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     color v1 -> %rcx
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring v2 = iconst.i64 2
    from [ GPR: --db--s-89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     glob [ GPR: acdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     color v2 -> %rdx
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring v3 = iconst.i64 3
    from [ GPR: ---b--s-89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     glob [ GPR: acdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     color v3 -> %rbx
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring store.i64 notrap aligned v0, v4
    from [ GPR: ------s-89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     glob [ GPR: acdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring store.i64 notrap aligned v1, v4+8
    from [ GPR: ------s-89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     glob [ GPR: acdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring store.i64 notrap aligned v2, v4+16
    from [ GPR: ------s-89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     glob [ GPR: acdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring store.i64 notrap aligned v3, v4+24
    from [ GPR: ------s-89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   >     glob [ GPR: acdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::coloring   > Coloring return v0, v1, v2, v3, v4
    from [ GPR: ------s-89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::solver     > reassign_in(v4:GPR, %rdi -> %rax)
 DEBUG cranelift_codegen::regalloc::solver     > add_var(v0:GPR, from=%rax)
 DEBUG cranelift_codegen::regalloc::solver     > -> new var: v0(GPR, from %rax, in, out)
 DEBUG cranelift_codegen::regalloc::coloring   >     kill v4 in %rdi (local GPR)
 DEBUG cranelift_codegen::regalloc::coloring   >     kill v0 in %rax (local GPR)
 DEBUG cranelift_codegen::regalloc::coloring   >     kill v1 in %rcx (local GPR)
 DEBUG cranelift_codegen::regalloc::coloring   >     kill v2 in %rdx (local GPR)
 DEBUG cranelift_codegen::regalloc::coloring   >     kill v3 in %rbx (local GPR)
 DEBUG cranelift_codegen::regalloc::coloring   >     glob [ GPR: acdb--sd89012345 FPR: 0123456789012345 FLAG: f ]
 DEBUG cranelift_codegen::regalloc::solver     > collect_moves: [v0:GPR(%rax -> %rsi), v4:GPR(%rdi -> %rax)]
 DEBUG cranelift_codegen::regalloc::solver     > move #0: v0:GPR(%rax -> %rsi)
 DEBUG cranelift_codegen::regalloc::solver     > move #1: v4:GPR(%rdi -> %rax)
 DEBUG cranelift_codegen::timing::details      > timing: Ending RA coloring
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify Cranelift IR, (during Register allocation)
 DEBUG cranelift_codegen::timing::details      > timing: Starting Control flow graph, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Control flow graph
 DEBUG cranelift_codegen::timing::details      > timing: Starting Dominator tree, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Dominator tree
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify CPU flags, (during Verify Cranelift IR)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify CPU flags
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify Cranelift IR
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify live ranges, (during Register allocation)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify live ranges
 DEBUG cranelift_codegen::timing::details      > timing: Starting Verify value locations, (during Register allocation)
 DEBUG cranelift_codegen::timing::details      > timing: Ending Verify value locations
 DEBUG cranelift_codegen::timing::details      > timing: Ending Register allocation
 DEBUG cranelift_codegen::timing::details      > timing: Ending Compilation passes
function %return_4_i64s(i64 sret [%rdi]) -> i64 [sret(0x0)], i64 [sret(0x8)], i64 [sret(0x10)], i64 [sret(0x18)], i64 sret [%rax] fast {
                                ebb0(v4: i64 [%rdi]):
[RexOp1pu_id#b8,%rax]               v0 = iconst.i64 0
[RexOp1pu_id#b8,%rcx]               v1 = iconst.i64 1
[RexOp1pu_id#b8,%rdx]               v2 = iconst.i64 2
[RexOp1pu_id#b8,%rbx]               v3 = iconst.i64 3
[RexOp1st#8089]                     store notrap aligned v0, v4
[RexOp1stDisp8#8089]                store notrap aligned v1, v4+8
[RexOp1stDisp8#8089]                store notrap aligned v2, v4+16
[RexOp1stDisp8#8089]                store notrap aligned v3, v4+24
[RexOp1rmov#8089]                   regmove v0, %rax -> %rsi
[RexOp1rmov#8089]                   regmove v4, %rdi -> %rax
[Op1ret#c3]                         return v0, v1, v2, v3, v4
;~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
; error: inst4: ABI expects v0 at return pointer offset 0, got %rsi

}

; 1 verifier error detected (see above). Compilation aborted.
```

</details>

----------

One idea I was considering (but don't actually know if this is a good idea or just a kludgy workaround) was introducing a `sret_store` (or `store_copy`) instruction that would return a new copy of the value that is located at the stored location (similar to `copy` and `spill`).

Basically, this would turn the previous example into:

```
function %return_4_i64s(i64 sret [%rdi]) -> i64 [sret(0x0)], i64 [sret(0x8)], i64 [sret(0x10)], i64 [sret(0x18)], i64 sret [%rax] fast {
ebb0(v4: i64):
    v0 = iconst.i64 0
    v1 = iconst.i64 1
    v2 = iconst.i64 2
    v3 = iconst.i64 3
    v5 = sret_store v4+0, v0
    v6 = sret_store v4+8, v0
    v7 = sret_store v4+16, v0
    v8 = sret_store v4+24, v0
    return v5, v6, v7, v8, v4
```

But I didn't really want to pursue this further without getting feedback from y'all.

-----------------

Thanks!

bjorn3 commented:
When a value is stored into a stack slot, it is normally done using `spill`, not `store`. The former says that the value moved, while the later stores a copy.

bnjbvr commented:
These two lines in the log:

```
 DEBUG cranelift_codegen::regalloc::solver     > reassign_in(v4:GPR, %rdi -> %rax)
 DEBUG cranelift_codegen::regalloc::solver     > add_var(v0:GPR, from=%rax)
```

mean that the solver (during the coloring phase of regalloc) is making room so that v4 can be in rax, which is the ABI return register (when only one value is returned); then v0 (which already lives in rax) must be diverted, so it must be added as a solver variable so rax is available for v4.

Is it possible instead to change the ABI location requirements of return values in `cranelift-codegen/src/isa/x86/abi.rs` when a function returns multiple values? Maybe even create the stack slot for the returned data structure's base there.

fitzgen commented:
> then v0 (which already lives in rax) must be diverted, so it must be added as a solver variable so rax is available for v4.

But the thing is `v0` *shouldn't* be considered to live in `%rax`, since it got stored into its sret slot and I updated the `ValueLocation` to reflect that. So it should be fine to overwrite `%rax` rather than try and divert its current value.

Basically, the problem is that I don't know why (a) updating the `ValueLocation` isn't enough to convince the register allocator of that, and (b) how to fix this (do I need to do more than update the `ValueLocation`? does the register allocator need to be fixed?).

> Is it possible instead to change the ABI location requirements of return values in `cranelift-codegen/src/isa/x86/abi.rs` when a function returns multiple values? Maybe even create the stack slot for the returned data structure's base there.

I can look into it / thought I was doing that by setting the values' `ArgumentLoc` to the new `RetPtrOffset` variant.

Regarding specifically creating the stack slot when looking at return values, I don't think that can work since the stack slot is in the caller's frame, not the callee's frame.

bjorn3 commented:
> Basically, the problem is that I don't know why (a) updating the ValueLocation isn't enough to convince the register allocator of that, and (b) how to fix this (do I need to do more than update the ValueLocation? does the register allocator need to be fixed?).

`[RexOp1pu_id#b8,sret(0)]            v0 = iconst.i64 0` is impossible. There is no encoding of `iconst` which stores its result at a certain memory location. It only supports storing in a register.

> One idea I was considering (but don't actually know if this is a good idea or just a kludgy workaround) was introducing a sret_store (or store_copy) instruction that would return a new copy of the value that is located at the stored location (similar to copy and spill).

I think this is the proper solution, as you would need a separate machine instruction to store the result anyway in many cases and every machine instruction corresponds to a clif instruction. (While the conditional trap instructions map to multiple machine instructions, they are an exceptional case.)

iximeow commented on cranelift-codegen/src/stack_layout.rs:
```diff
@@ -56,6 +58,14 @@ pub fn layout_stack(frame: &mut StackSlots, alignment: StackSize) -> CodegenResu
                     .ok_or(CodegenError::ImplLimitExceeded)?;
                 outgoing_max = max(outgoing_max, offset);
             }
+            StackSlotKind::RetPtr => {
```
This feels like a kind of silly question: where are `StackSlot` with kind `RetPtr` added? I'd expect iterating all calls in a function and adding the necessary slots from that, but don't see anything of the sort.

edit: I think I see the relevant `TODO` in legalizer/boundary.rs :)

iximeow replied:
As a followup, and I think there ought to be a test case (or plural!) around this, if we're reusing the largest `RetPtr` slot for different multi-return calls with spills, I'd like to be sure that we know this stack slot is clobbered on reuse. That way Cranelift can spill out of this slot into somewhere that return values can be safely read from. eg `v0` should be moved out of the stack-returns area in this clif, which I think is pretty close to valid:
```
function %retptr_reuse() fast {
    sig0 = () -> i32, i32, i32, i32, i32, i32 fast
    sig1 = () -> f32, f32, f32, f32, f32, f32 fast

    ireturns = u0:0 sig0
    freturns = u0:1 sig1

ebb0():
    v0, v1, v2, v3, v4, v5 = call icalls()
    v6, v7, v8, v9, v10, v11 = call fcalls()
    ; I'm concerned v6 will overwrite v0, as things are currently
}
```

I don't know if the notion of stack slots being reused for different purposes is even something Cranelift knows about right know?

fitzgen replied:
I haven't really started on implementing the caller side yet; sorry I should have been more clear in the op.

Yeah I think it is worth not attempting to reuse these stack slots at all in the first pass implementation, and then circling back to add the reuse as a new PR.

iximeow commented on cranelift-codegen/src/legalizer/boundary.rs:
```diff
@@ -386,27 +394,46 @@ fn convert_to_abi<PutArg>(
-    let (sig_ref, args) = match dfg[inst].analyze_call(&dfg.value_lists) {
-        CallInfo::Direct(func, args) => (dfg.ext_funcs[func].signature, args),
+    let (sig_ref, args) = match func.dfg[inst].analyze_call(&func.dfg.value_lists) {
+        CallInfo::Direct(f, args) => (func.dfg.ext_funcs[f].signature, args),
```
small name comment, `target` or `dest` instead of `f`?

iximeow commented on cranelift-codegen/src/isa/x86/abi.rs:
```diff
@@ -205,23 +210,74 @@ pub fn legalize_signature(
+                ret_ptr_param.location = loc;
+                sig.params.insert(0, ret_ptr_param);
+            }
+            ArgAction::Convert(_) => unreachable!(),
```
Is it correct to phrase this as ``unreachable!("`pointer_type` shouldn't require conversion")``?

bnjbvr commented:
> Any idea what might be going on and what I should do from here?

Ok, I re-read it and read your code too. v0 is not dead, since it's referenced in the return instruction. Does it work when you only keep the last value (i.e. the ret struct point) in the IR return instruction, when legalizing the signature? There's apparently no need to keep the other values live in the ret instruction, since they're manually moved to their final location before the actual ret.

fitzgen commented:
Thanks for the feedback and tips @bjorn3, @iximeow, and @bnjbvr!

> Does it work when you only keep the last value (i.e. the ret struct point) in the IR return instruction, when legalizing the signature? There's apparently no need to keep the other values live in the ret instruction, since they're manually moved to their final location before the actual ret.

I think this actually makes more sense than the `sret_store` instruction, since we need loads on the caller side as well, and if we have return values that don't really "exist" until loaded that seems funky given my new/better understanding of what's going on here.

I'll try legalizing this signature

```
function %return_4_i64s() -> i64, i64, i64, i64 fast
```

into

```
function %return_4_i64s(v0: i64 sret) -> i64 sret fast
```

and legalizing returns into stores, as before, except with only the single return retptr value

```
;; before
ebb0:
    return v0, v1, v2, v3

;; after
ebb0(v4: i64):
    v0 = iconst.i64 0
    v1 = iconst.i64 1
    v2 = iconst.i64 2
    v3 = iconst.i64 3
    store notrap aligned v0, v4
    store notrap aligned v1, v4+8
    store notrap aligned v2, v4+16
    store notrap aligned v3, v4+24
    return v4
```

and finally legalizing calls like this

```
;; before
ebb0:
    v0, v1, v2, v3 = call %return_4_i64s

;; after
    ss0 = ret_ptr 32
ebb0:
    v4 = stack_addr.i64 ss0
    v5 = call %return_4_i64s
    v0 = stack_load.i64 ss0
    v1 = stack_load.i64 ss0+8
    v2 = stack_load.i64 ss0+16
    v3 = stack_load.i64 ss0+24
```

This feels very promising and I think I can actually remove a lot of the extra `ArgumentLoc` and `ValueLoc` stuff that I added.

Looking back, I don't know why I was so determined to keep the multiple return values in the signature, instead of letting them dissolve away into the loads and stores like they actually do at the machine level...

fitzgen commented on cranelift-codegen/src/stack_layout.rs:
```diff
@@ -97,6 +108,14 @@ pub fn layout_stack(frame: &mut StackSlots, alignment: StackSize) -> CodegenResu
+    // Make room for the maximum amount of incoming return pointer space we'll
+    // need.
+    //
+    // TODO FITZGEN: alignment?
```
This is regarding the alignment required by the return values. For example, if for some reason there is an `i32` stack slot, and then we need another stack slot for 8-byte-aligned data, we would need to add padding after the `i32` slot (or rearrange them) and I'm just not sure yet how much of this infra already exists and is done for me automatically or not, and how to express these requirements if it does exist.

bjorn3 commented:
> Looking back, I don't know why I was so determined to keep the multiple return values in the signature, instead of letting them dissolve away into the loads and stores like they actually do at the machine level...

It could help with a consumer generating debuginfo for return values from the post compilation ir.

fitzgen commented:
Ok! I have all tests passing plus a bunch of new tests specifically for multi-value stuff :)

I think this is ready for review! r? @bnjbvr

fitzgen commented:
@bnjbvr 

> This one is trickier, because i'm not sure we have enough runtime support in the `run` test framework for this (in particular with respect to linking), but: it would be extra nice to have `run` tests for functions returning multiple values and checking that the returned values are correct.

Yeah, the `test run` mode doesn't support calls, since it compiles each function individually. We have calls [working in wasmtime](https://github.com/CraneStation/wasmtime/pull/467) and tests that they work right, so would you be OK with skipping `test run` tests here?

fitzgen commented:
> Have you considered de-grouping return values, that is, instead of having a single `RetPtr` stack slot, have several `OutgoingRet` stack slots? I wonder if this wouldn't make a few things simpler, like, no need to precompute the size of the entire ret stack space, so the two big loops when legalizing a call instruction could be fused.

In order to pass in a single `sret` pointer, we need to ensure that the space for each of these returns are contiguous and that they are each aligned. I think upholding those properties would only be harder / less clear that they *are* being upheld if we tried to explode the `ret_ptr` stack slots into multiple pieces.

fitzgen commented on cranelift-codegen/src/isa/x86/abi.rs:
```diff
@@ -165,6 +165,89 @@ impl ArgAssigner for Args {
+                            | ArgAction::Convert(ValueConversion::IntBits)
+                            | ArgAction::Convert(ValueConversion::Sext(_))
+                            | ArgAction::Convert(ValueConversion::Uext(_)) => {
+                                continue 'draining_assignments;
```
I find that this is easier to read and understand when there are nested loops, just so that everyone is clear on what control flow is doing and doesn't wonder "did this happen to move and change which thing is being `continue`d or `break`ed?" but if you feel that it doesn't help code clarity I can remove it.

bnjbvr replied:
For me, seeing these labels rings a bell and tells me that something weird/complex is going on: when looking at this code, when I saw the labels, I imagined that an innermost loop would break out to the outermost loop. It wasn't the case, so I figured these labels were perhaps here to ease reading. So I was looking for complexity that didn't exist, triggered by the loop header labels. I would be in favor of removing them because 1. they may introduce confusion for readers, 2. minimalism in principle. Since it's highly subjective, I'll let you make the final call.

fitzgen replied:
Ok, I can get rid of them, because that's not the reaction I want from code readers ;)

fitzgen commented on cranelift-codegen/src/isa/x86/abi.rs:
```diff
@@ -165,6 +165,89 @@ impl ArgAssigner for Args {
+    isa_flags: &isa_settings::Flags,
+    params: impl IntoIterator<Item = &'a AbiParam>,
+) -> (usize, usize) {
+    let gprs = &[RU::rax; 128];
```
The idea is that we are using an "infinite" number of registers here and counting how many get used in this ideal, infinite registers world, and then later we are comparing that count with how many we *actually* have available (which does take into account the ABI-specific counts).

I can add a clarifying comment to this effect.

fitzgen commented on cranelift-codegen/src/isa/x86/abi.rs:
```diff
@@ -165,6 +165,89 @@ impl ArgAssigner for Args {
+        let mut split_factor = 1;
+
+        'converting_param: loop {
+            match assigner.assign(&p) {
```
Non-normal purpose parameters will still use registers which can affect this calculation, so skipping them would open the door for bugs. I do think that filtering out already-assigned parameters is a good idea.

fitzgen commented on cranelift-codegen/src/isa/x86/abi.rs:
```diff
@@ -165,6 +165,89 @@ impl ArgAssigner for Args {
+                    p.value_type = p.value_type.half_vector().unwrap();
+                }
+                ArgAction::Assign(ArgumentLoc::Reg(_))
+                | ArgAction::Convert(ValueConversion::IntBits)
```
That would be nice, excpet `legalize_args` mutates the parameters, and we might need to undo those mutations if we decide to use a struct return pointer instead of registers. For example, `legalize_args` might split values into smaller pieces with less alignment requirements, that we then need to reconstruct and ensure are aligned if we store it through the struct return pointer.

The easiest way to do that would be to keep a clone of the vec of original parameters that we can restore if necessary, but that clone has a potential perf implication.

Would you prefer to do that clone rather than count the registers required?

Sort of an aside: assigning parameters and returns to registers would really benefit from some re-working to make backtracking easier, as sys v (and I'm sure other calling ocnventions too) need that in various scenarios... :-/

fitzgen commented on cranelift-codegen/src/isa/x86/abi.rs:
```diff
@@ -206,26 +289,77 @@ pub fn legalize_signature(
+    } {
+        debug_assert!(!sig.uses_sret());
+
+        // We're using the first register for the return pointer parameter.
```
This PR is only growing some support for struct returns. It isn't changing anything that already worked, and anyone trying to return multiple values that would be affected by this PR would previously be getting validation errors.

In particular, I took care to make sure that multi-value returns that *did* work by happenstance continue to work, e.g. returning a pair of `i64`s in registers.

It is true that full support for sys v struct returns (which requires understanding structs, padding, and ctors/dtors) and whatever other calling conventions will require more work than this PR is providing, but those things already don't work today, and this is getting us closer in an incremental fashion. This is really just for our unstable, internal "fast" calling convention and everything else that already worked is left alone. Things that didn't work before might happen to do *something* now by happenstance for other calling conventions, but this doesn't seem like the end of the world. We can always revisit asserting that multi-value is only used with calling conventions that we know we implement multi-value support for properly (e.g. only our internal, unstable "fast" calling convention) but these seemed undesirable for the rustc cranelift backend, which relies on the happenstance support right now.

fitzgen commented on cranelift-codegen/src/legalizer/boundary.rs:
```diff
@@ -528,6 +681,67 @@ fn legalize_inst_arguments<ArgType>(
+                let mut new_ty = new_ty;
+
+                if ty.is_bool() {
+                    new_ty = new_ty.as_bool_pedantic();
```
This is converting `new_ty` from an integer type to a boolean type, which I would expect to be a no-op at the machine level, but legalization / instruction selection isn't happy later on if we didn't do this conversion.

fitzgen commented on cranelift-codegen/src/legalizer/boundary.rs:
```diff
@@ -528,6 +681,67 @@ fn legalize_inst_arguments<ArgType>(
+
+                if ty.bits() < new_ty.bits() {
+                    new_ty = ty;
+                    v = pos.ins().breduce(new_ty, v);
```
This case will only be hit with boolean types, but I can move it into the other `if` to clarify that.

fitzgen commented on cranelift-codegen/src/legalizer/boundary.rs:
```diff
@@ -265,6 +265,152 @@ where
+    let sret_arg = pos.ins().stack_addr(ptr_type, stack_slot, 0);
+    let mut args = pos.func.dfg[call].take_value_list().unwrap();
+    args.push(sret_arg, &mut pos.func.dfg.value_lists);
+    pos.func.dfg[call].put_value_list(args);
```
I thought there wasn't, but upon going to add one, it turns out there is!

fitzgen commented on cranelift-codegen/src/legalizer/boundary.rs:
```diff
@@ -265,6 +265,152 @@ where
+    let stack_offset = pos.func.stack_slots.values().fold(0, |off, slot| {
+        let slot_size = slot.size as i32;
+        let slot_offset = slot.offset.unwrap_or(-slot_size);
+        std::cmp::max(off, (slot_offset + slot_size) as u32)
```
I found that they have been, and that if I set the offset to `None`, then things would break later on because they assumed it had already been set, but I didn't know whether I should assert that the offsets are already set here or not.

bnjbvr replied:
I've thought about it: this function will be called during legalization, so well before register allocation. The latter might introduce spill slots, and since the ordering (as described in `stack_layout.rs`) puts the explicit slots / spill slots *before* the incoming return stack space, I wonder if a spill after a call to a multi-return function would cause an incorrect offset here (leading to incorrect behavior, probably stack clobbering)? If so, I imagine the fix would be to put the struct stack space *before* the explicit / spill slots, and ignore stack slots of kinds appearing after the stack-ret-space in this offset computation.

fitzgen replied:
Ok, I've switched things up a little and removed a bunch of code, and now struct return slots are not given pre-assigned offsets in the frame, instead they are assigned offsets by the existing frame layout algorithm. Nice little clean up, thanks for pushing back on this!

fitzgen commented on cranelift-codegen/src/legalizer/boundary.rs:
```diff
@@ -644,10 +870,39 @@ pub fn handle_return_abi(inst: Inst, func: &mut Function, cfg: &ControlFlowGraph
             vlist.push(val, &mut pos.func.dfg.value_lists);
+
+            if let ArgumentPurpose::StructReturn = arg.purpose {
+                sret = Some(val);
```
Unless I am overlooking something, I don't think there is an easy way to do it in the `match` above, because we don't have the `val` inserted in the value list at that point, and we would have to duplicate almost all the rest of this loop's body.

bnjbvr replied:
You're right, sorry for the noise.

fitzgen commented:
Ok, I think I've addressed all of your feedback @bnjbvr, other than the comments that I replied to inline above. This should be ready for a second round of review. I tried to make each thing you requested into its own commit so that it is easier for you to review. Thanks!

bnjbvr commented:
Thanks for the answers and commits. I'll take a look at this on Monday, I'd like to re-read it entirely.

Just one question that came up when discussing this with people: do we really need the pointer-to-stack? I seem to recall that stack space is allocated at once, and the callee could know statically where to put the return values, with a fixed convention (that we have with this patch). I don't want to block your work on this, just wanted to highlight an potential opportunity for a follow-up here.

bjorn3 commented:
> Just one question that came up when discussing this with people: do we really need the pointer-to-stack?

If we want to follow the System-V call conv, we will need to pass a return value address like this PR does. That value is passes in %rdi and the same pointer should be returned in %rax.

fitzgen commented:
> Just one question that came up when discussing this with people: do we really need the pointer-to-stack? I seem to recall that stack space is allocated at once, and the callee could know statically where to put the return values, with a fixed convention (that we have with this patch).

Do you mean like statically allocating space for a function's return values in a single place somewhere, rather than in every stack frame? Re-entrancy shouldn't be a problem, at least right now, since we immediately load the values out of the sret space on return. One could theoretically even not pass the pointer to the space in or out of the function, if doing this. But that is only going to work for internal function calls, and won't work for external, cross-shared library calls. And of course, it isn't what sys v (or any other calling convention I know of) does, as @bjorn3 mentioned, so we would still need support for alternative mechanisms.

Interesting idea though!

bnjbvr commented:
> Do you mean like statically allocating space for a function's return values in a single place somewhere, rather than in every stack frame?

No, I was rather imagining that we could allocate the stack space in the caller, close to the outgoing parameters area. Then the callee statically knows where the incoming arguments are in this area, i.e. it can compute static statck offsets to each argument. The same way, the callee could compute the static offset to the caller return stack space.

As pointed out by @bjorn3, this wouldn't support calling out to native functions with SysV. Note that support for Spidermonkey might need an alternative implementation strategy later, so having a way to choose among different possible implementations will happen at some point. But, later :)

fitzgen commented:
> No, I was rather imagining that we could allocate the stack space in the caller, close to the outgoing parameters area. Then the callee statically knows where the incoming arguments are in this area, i.e. it can compute static statck offsets to each argument. The same way, the callee could compute the static offset to the caller return stack space.

Ah I see. Yes, this is another possibility to explore at some point in the future!

bnjbvr commented on cranelift-codegen/src/legalizer/boundary.rs:
```diff
@@ -245,6 +266,176 @@ where
     call
 }
 
+fn assert_is_valid_sret_legalization(
```
Should the function definition be guarded by a cfg guard too, to avoid dead code warnings in release mode?

fitzgen replied:
Because the call is inside a `if cfg!(...) { assert_blah(); }` it will expand to `if false { assert_blah(); }` which won't trigger dead code warnings, but will get DCE'd.

If instead of `if cfg!(...) { ... }`, we did

```rust
#[cfg(debug_assertions)]
{
    ...
}
```

then the block would expand into nothing, and there would be no caller of the assertion function, and it would trigger a dead code warning in release builds.

bnjbvr commented on cranelift-codegen/src/legalizer/boundary.rs:
```diff
@@ -639,6 +914,56 @@ pub fn handle_return_abi(inst: Inst, func: &mut Function, cfg: &ControlFlowGraph
+    //     (n + to - 1) / to * to
+    //
+    // Consider the numerator: `n + to - 1`. This is ensuring that if there is
+    // any remainder for `n / to`, than the result of the division is one
```
nit: `than` => `then`

bnjbvr commented on cranelift-codegen/src/legalizer/boundary.rs:
```diff
@@ -714,3 +1039,34 @@ fn spill_call_arguments(pos: &mut FuncCursor) -> bool {
+            (5, 4, 8),
+        ] {
+            let actual = round_up_to_multiple_of_pow2(n, to);
+            println!(
```
Can the message be folded into the `assert_eq` message instead?

bnjbvr commented on cranelift-codegen/src/legalizer/boundary.rs:
```diff
@@ -245,6 +266,176 @@ where
+        let mut new_v = new_legalized_v;
+        if ty.is_bool() {
+            legalized_ty = legalized_ty.as_bool_pedantic();
+            new_v = pos.ins().raw_bitcast(legalized_ty, new_v);
```
Out of curiosity, can't the above `load` handle boolean values? (can be done as a follow up otherwise, since it might require a few new encodings)

fitzgen replied:
If we had the encodings, yes I believe it could be.

bnjbvr commented on filetests/wasm/multi-val-reuse-ret-ptr-stack-slot.clif:
```diff
@@ -0,0 +1,61 @@
+
+;; Test that we don't reuse `sret` stack slots for multiple calls. We could do
+;; this one day, but it would require some care to ensure that we don't have
+;; subsequent calls overwrite the results of previous calls.
```
I'm assuming this is working as intended, but... isn't the max'ing happening in the `layout_stack` function all about trying to reuse the same slot? Could we also check the actual amount of stack pushed for the stack return space?
